pkg/manager-nvme/manager.go

```diff
@@ -404,6 +432,28 @@ func (s *Storage) purge() error {
 	return nil
 }
 
+func (s *Storage) purgeVolumes(volIdsToKeep []string) error {
+	if s.device == nil {
+		return fmt.Errorf("Storage %s has no device", s.id)
+	}
+
+vol_loop:
+	for _, vol := range s.volumes {
+		for _, volIdToKeep := range volIdsToKeep {
+			if volIdToKeep == vol.id {
+				continue vol_loop
+			}
+		}
+
+		log.Infof("Storage %s Volume ID %s is being removed", s.id, vol.id)
+		if err := s.deleteVolume(vol.id); err != nil {
```

Comment:
`s.deleteVolume()` modifies the `s.volumes` array. So your iterator on line 441 will be corrupted by the call to `deleteVolume()`

This is a common gotcha in the nnf-ec code. It's almost like nnf-ec code was written by a complete beginner.

Reply:
I updated it to collect the list of vol ids to delete. Then, once we're done iterating `s.volumes`, each volume is deleted.